OrbitVulkanLayer/DispatchTable.h

```diff
@@ -0,0 +1,359 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_VULKAN_LAYER_DISPATCH_TABLE_H_
+#define ORBIT_VULKAN_LAYER_DISPATCH_TABLE_H_
+
+#include "OrbitBase/Logging.h"
+#include "absl/base/casts.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "vulkan/vk_layer.h"
+#include "vulkan/vk_layer_dispatch_table.h"
+#include "vulkan/vulkan.h"
+
+namespace orbit_vulkan_layer {
+
+/*
+ * A thread-safe dispatch table for Vulkan function look-up.
+ *
+ * It computes/stores the Vulkan dispatch tables for concrete devices/instances and provides
+ * accessors to the functions.
+ *
+ * For functions provided by extensions it also provides predicate functions to check if the
+ * extension is available.
+ *
+ * Thread-Safety: This class is internally synchronized (using read/write locks) and can be safely
+ * accessed from different threads.
+ */
+class DispatchTable {
+ public:
+  DispatchTable() = default;
+
+  void CreateInstanceDispatchTable(VkInstance instance,
+                                   PFN_vkGetInstanceProcAddr next_get_instance_proc_addr_function);
+  void RemoveInstanceDispatchTable(VkInstance instance);
+
+  void CreateDeviceDispatchTable(VkDevice device,
+                                 PFN_vkGetDeviceProcAddr next_get_device_proc_addr_function);
+  void RemoveDeviceDispatchTable(VkDevice device);
+
+  template <typename DispatchableType>
+  PFN_vkDestroyDevice DestroyDevice(DispatchableType dispatchable_object) {
+    void* key = GetDispatchTableKey(dispatchable_object);
+    {
+      absl::ReaderMutexLock lock(&mutex_);
+      CHECK(device_dispatch_table_.contains(key));
+      CHECK(device_dispatch_table_.at(key).DestroyDevice != nullptr);
+      return device_dispatch_table_.at(key).DestroyDevice;
+    }
+  }
+
+  template <typename DispatchableType>
+  PFN_vkDestroyInstance DestroyInstance(DispatchableType dispatchable_object) {
+    void* key = GetDispatchTableKey(dispatchable_object);
+    {
+      absl::ReaderMutexLock lock(&mutex_);
+      CHECK(instance_dispatch_table_.contains(key));
```

Comment:
Are these methods called on every call to whatever function is being dispatched? How much overhead do all these CHECKS generate? If these methods are only called once for some form of initialization, it doesn't matter.

Reply:
They are called on each dispatch. So yes, overhead might be a problem. I haven't done any proper performance analysis of the layer. So far it does not seem to be an issue.
I'd suggest for now to keep it as it is, and if we have a proper analysis of the overhead, we can check if e.g. DCHECK is reasonable.

Reply:
Ok, we'll have to take a close look at overhead anyway. Let's keep them for now and change if needed later.